Source/Core/DolphinQt/GCMemcardManager.cpp

```diff
@@ -432,94 +431,94 @@ void GCMemcardManager::FixChecksums()
 
 void GCMemcardManager::DrawIcons()
 {
-  m_current_frame++;
-  m_current_frame %= 15;
-
   const auto column = 3;
   for (int slot = 0; slot < SLOT_COUNT; slot++)
   {
     int row = 0;
-    for (auto& icon : m_slot_active_icons[slot])
+    for (const auto& icon : m_slot_active_icons[slot])
     {
-      int frame = (m_current_frame / 3 - icon.first) % icon.second.size();
+      const u64 current_time_in_animation = m_current_frame % icon.m_frame_timing.size();
+      const u8 current_frame = icon.m_frame_timing[current_time_in_animation];
 
       auto* item = new QTableWidgetItem;
-      item->setData(Qt::DecorationRole, icon.second[frame]);
+      item->setData(Qt::DecorationRole, icon.m_frames[current_frame]);
       item->setFlags(item->flags() ^ Qt::ItemIsEditable);
 
       m_slot_table[slot]->setItem(row, column, item);
       row++;
     }
   }
+
+  ++m_current_frame;
 }
 
 QPixmap GCMemcardManager::GetBannerFromSaveFile(int file_index, int slot)
 {
   auto& memcard = m_slot_memcard[slot];
 
-  std::vector<u32> pxdata(BANNER_WIDTH * IMAGE_HEIGHT);
+  std::vector<u32> pxdata(BANNER_WIDTH * BANNER_HEIGHT);
 
   QImage image;
   if (memcard->ReadBannerRGBA8(file_index, pxdata.data()))
   {
-    image = QImage(reinterpret_cast<u8*>(pxdata.data()), BANNER_WIDTH, IMAGE_HEIGHT,
+    image = QImage(reinterpret_cast<u8*>(pxdata.data()), BANNER_WIDTH, BANNER_HEIGHT,
                    QImage::Format_ARGB32);
   }
 
   return QPixmap::fromImage(image);
 }
 
-std::vector<QPixmap> GCMemcardManager::GetIconFromSaveFile(int file_index, int slot)
+GCMemcardManagerIconAnimationData GCMemcardManager::GetIconFromSaveFile(int file_index, int slot)
 {
   auto& memcard = m_slot_memcard[slot];
 
-  std::vector<u32> pxdata(BANNER_WIDTH * IMAGE_HEIGHT);
   std::vector<u8> anim_delay(ANIM_MAX_FRAMES);
-  std::vector<u32> anim_data(ANIM_FRAME_WIDTH * IMAGE_HEIGHT * ANIM_MAX_FRAMES);
+  std::vector<u32> anim_data(ICON_WIDTH * ICON_HEIGHT * ANIM_MAX_FRAMES);
 
-  std::vector<QPixmap> frame_pixmaps;
+  GCMemcardManagerIconAnimationData frame_data;
 
   u32 num_frames = memcard->ReadAnimRGBA8(file_index, anim_data.data(), anim_delay.data());
+  bool is_pingpong = memcard->DEntry_IsPingPong(file_index);
```

Comment:
This can be moved into the `num_frames > 0` conditional scope below and be made const.